Unable to compile when dynamically linking `wasmtime` crate.

It's known dynamically linking dependencies in large Rust projects can improve compile times to a considerable degree. Facing the following problem when attempting to setup such a workspace:
```
wasmtime_dylink_issue on  main [?] is 📦 v0.1.0 via 🦀 v1.54.0-nightly took 7s
❯ cargo r
    .... <loads of compiled crates>...
   Compiling deps v0.1.0 (/home/user/Source/rust/wasmtime_dylink_issue/crates/deps)
   Compiling dylink v0.1.0 (/home/user/Source/rust/wasmtime_dylink_issue/crates/dylink)
   Compiling wasmtime_dylink_issue v0.1.0 (/home/user/Source/rust/wasmtime_dylink_issue)

error: linking with `cc` failed: exit status: 1
  |
  = note: "cc" "-m64" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-Wl,--as-needed" "-L" "/home/user/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/home/user/Source/rust/wasmtime_dylink_issue/target/debug/deps/wasmtime_dylink_issue-223c9046feaa7cea.176re6pgxyy8ex5.rcgu.o" ......<loads of object files>...... "/home/user/Source/rust/wasmtime_dylink_issue/target/debug/deps/wasmtime_dylink_issue-223c9046feaa7cea.yqhnz0r7qoxxab5.rcgu.o" "-o" "/home/user/Source/rust/wasmtime_dylink_issue/target/debug/deps/wasmtime_dylink_issue-223c9046feaa7cea" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-nodefaultlibs" "-L" "/home/user/Source/rust/wasmtime_dylink_issue/target/debug/deps" "-L" "/home/user/Source/rust/wasmtime_dylink_issue/target/debug/build/psm-04165501c77b945c/out" "-L" "/home/user/Source/rust/wasmtime_dylink_issue/target/debug/build/zstd-sys-0f0a0a67e9ce9bc7/out" "-L" "/home/user/Source/rust/wasmtime_dylink_issue/target/debug/build/wasmtime-fiber-8ff9cbc1eca6d35f/out" "-L" "/home/user/Source/rust/wasmtime_dylink_issue/target/debug/build/wasmtime-runtime-2bce6ad7a15d697a/out" "-L" "/home/user/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/home/user/Source/rust/wasmtime_dylink_issue/target/debug/deps" "-ldylink" "-Wl,--start-group" "-L" "/home/user/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-lstd-c6dddd3d354e6bea" "-Wl,--end-group" "-Wl,-Bstatic" "/home/user/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-8b33f9cbbc9652fe.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc"
  = note: /bin/ld: /home/user/Source/rust/wasmtime_dylink_issue/target/debug/deps/wasmtime_dylink_issue223c9046feaa7cea.3nou4ari0sahe2y2.rcgu.o: in function `wasmtime_runtime::traphandlers::catch_traps::{{closure}}':
          /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/wasmtime-runtime-0.28.0/src/traphandlers.rs:180: undefined reference to `RegisterSetjmp'
          collect2: error: ld returned 1 exit status


error: aborting due to previous error; 1 warning emitted

error: could not compile `wasmtime_dylink_issue`

To learn more, run the command again with --verbose.
```

[Repo to recreate bug.](https://github.com/skorpi-and-friends/wasmtime_dylink_issue)

Similar issue when using the llvm linker.

Not exactly a traditional bug, I know, but would really appreciate any recs for getting past this.

--- a/crates/runtime/build.rs
+++ b/crates/runtime/build.rs
@@ -9,5 +9,5 @@ fn main() {
             None,
         )
         .file("src/helpers.c")
-        .compile("helpers");
+        .compile("wasmtime-helpers");
 }

--- a/crates/runtime/src/helpers.c
+++ b/crates/runtime/src/helpers.c
@@ -14,7 +14,7 @@
 #define platform_jmp_buf sigjmp_buf
 #endif
 
-int RegisterSetjmp(
+int wasmtime_setjmp(
     void **buf_storage,
     void (*body)(void*, void*),
     void *payload,
@@ -28,7 +28,7 @@ int RegisterSetjmp(
   return 1;
 }
 
-void Unwind(void *JmpBuf) {
+void wasmtime_longjmp(void *JmpBuf) {
   platform_jmp_buf *buf = (platform_jmp_buf*) JmpBuf;
   platform_longjmp(*buf, 1);
 }

--- a/crates/runtime/src/traphandlers.rs
+++ b/crates/runtime/src/traphandlers.rs
@@ -14,15 +14,16 @@ use wasmtime_environ::ir;
 
 pub use self::tls::{tls_eager_initialize, TlsRestore};
 
+#[link(name = "wasmtime-helpers")]
 extern "C" {
     #[allow(improper_ctypes)]
-    fn RegisterSetjmp(
+    fn wasmtime_setjmp(
         jmp_buf: *mut *const u8,
         callback: extern "C" fn(*mut u8, *mut VMContext),
         payload: *mut u8,
         callee: *mut VMContext,
     ) -> i32;
-    fn Unwind(jmp_buf: *const u8) -> !;
+    fn wasmtime_longjmp(jmp_buf: *const u8) -> !;
 }
 
 cfg_if::cfg_if! {
@@ -177,7 +178,7 @@ where
     F: FnMut(*mut VMContext),
 {
     return CallThreadState::new(signal_handler).with(vminterrupts, |cx| {
-        RegisterSetjmp(
+        wasmtime_setjmp(
             cx.jmp_buf.as_ptr(),
             call_closure::<F>,
             &mut closure as *mut F as *mut u8,
@@ -251,7 +252,7 @@ impl CallThreadState {
     fn unwind_with(&self, reason: UnwindReason) -> ! {
         unsafe {
             (*self.unwind.get()).as_mut_ptr().write(reason);
-            Unwind(self.jmp_buf.get());
+            wasmtime_longjmp(self.jmp_buf.get());
         }
     }
 
@@ -306,7 +307,7 @@ impl CallThreadState {
         }
 
         // If all that passed then this is indeed a wasm trap, so return the
-        // `jmp_buf` passed to `Unwind` to resume.
+        // `jmp_buf` passed to `wasmtime_longjmp` to resume.
         self.jmp_buf.get()
     }
 

--- a/crates/runtime/src/traphandlers/macos.rs
+++ b/crates/runtime/src/traphandlers/macos.rs
@@ -33,7 +33,7 @@
 
 #![allow(non_snake_case)]
 
-use crate::traphandlers::{tls, Trap, Unwind};
+use crate::traphandlers::{tls, wasmtime_longjmp, Trap};
 use mach::exception_types::*;
 use mach::kern_return::*;
 use mach::mach_init::*;
@@ -389,7 +389,7 @@ unsafe extern "C" fn unwind(wasm_pc: *const u8) -> ! {
         state.jmp_buf.get()
     });
     debug_assert!(!jmp_buf.is_null());
-    Unwind(jmp_buf);
+    wasmtime_longjmp(jmp_buf);
 }
 
 thread_local! {

--- a/crates/runtime/src/traphandlers/unix.rs
+++ b/crates/runtime/src/traphandlers/unix.rs
@@ -1,4 +1,4 @@
-use crate::traphandlers::{tls, Trap, Unwind};
+use crate::traphandlers::{tls, wasmtime_longjmp, Trap};
 use std::cell::RefCell;
 use std::convert::TryInto;
 use std::io;
@@ -99,7 +99,7 @@ unsafe extern "C" fn trap_handler(
             return true;
         }
         info.capture_backtrace(pc);
-        Unwind(jmp_buf)
+        wasmtime_longjmp(jmp_buf)
     });
 
     if handled {

--- a/crates/runtime/src/traphandlers/windows.rs
+++ b/crates/runtime/src/traphandlers/windows.rs
@@ -1,4 +1,4 @@
-use crate::traphandlers::{tls, Trap, Unwind};
+use crate::traphandlers::{tls, wasmtime_longjmp, Trap};
 use std::io;
 use winapi::um::errhandlingapi::*;
 use winapi::um::minwinbase::*;
@@ -69,7 +69,7 @@ unsafe extern "system" fn exception_handler(exception_info: PEXCEPTION_POINTERS)
             EXCEPTION_CONTINUE_EXECUTION
         } else {
             info.capture_backtrace(ip);
-            Unwind(jmp_buf)
+            wasmtime_longjmp(jmp_buf)
         }
     })
 }

